workspaces/adventure-pack/goodies/typescript/Map.groupBy/index.test.ts

```diff
@@ -0,0 +1,189 @@
+    const result = Map.groupBy(set, (fruit) => fruit[0]);
+
+    expect(result).toStrictEqual(
+      new Map([
+        ["a", ["apple"]],
+        ["b", ["banana"]],
+        ["c", ["cherry"]],
+      ]),
+    );
+  });
+
+  it("should work with custom iterables", () => {
+    const customIterable = {
+      *[Symbol.iterator]() {
+        yield 1;
+        yield 2;
+        yield 3;
+      },
+    };
+    const result = Map.groupBy(customIterable, (num) =>
+      num % 2 === 0 ? "even" : "odd",
+    );
+
+    expect(result).toStrictEqual(
+      new Map([
+        ["even", [2]],
+        ["odd", [1, 3]],
+      ]),
+    );
+  });
+
+  it("should work with generators", () => {
+    function* numberGenerator() {
+      yield 1;
+      yield 2;
+      yield 3;
+    }
+    const result = Map.groupBy(numberGenerator(), (num) =>
+      num % 2 === 0 ? "even" : "odd",
+    );
+
+    expect(result).toStrictEqual(
+      new Map([
+        ["even", [2]],
+        ["odd", [1, 3]],
+      ]),
+    );
+  });
+
+  it("should work with Maps", () => {
+    const map = new Map([
+      ["a", 1],
+      ["b", 2],
+      ["c", 3],
+    ]);
+    const result = Map.groupBy(map, ([_, value]) =>
+      value % 2 === 0 ? "even" : "odd",
+    );
+
+    expect(result).toStrictEqual(
+      new Map([
+        ["even", [["b", 2]]],
+        [
+          "odd",
+          [
+            ["a", 1],
+            ["c", 3],
+          ],
+        ],
+      ]),
+    );
+  });
+
+  it("should handle objects with custom iterator", () => {
+    const obj: Iterable<number> = {
+      *[Symbol.iterator]() {
+        const data = [1, 2, 3, 4];
+        for (const item of data) {
+          yield item;
+        }
+      },
+    };
+    const result = Map.groupBy(obj, (num) => (num % 2 === 0 ? "even" : "odd"));
+
+    expect(result).toStrictEqual(
+      new Map([
+        ["even", [2, 4]],
+        ["odd", [1, 3]],
+      ]),
+    );
+  });
+
+  it("should throw TypeError for non-iterable arguments", () => {
+    expect(() => {
+      // @ts-ignore - Intentionally passing invalid type
+      Map.groupBy(123, (x) => x);
+    }).toThrow(TypeError);
+  });
+
+  it("should handle undefined or null keys", () => {
+    const data = [1, 2, 3, 4, 5];
+    const result = Map.groupBy(data, (num) =>
+      num % 2 === 0 ? undefined : null,
+    );
+
+    expect(result).toStrictEqual(
+      new Map([
+        [undefined, [2, 4]],
+        [null, [1, 3, 5]],
+      ]),
+    );
+  });
+
+  it("should handle keys that are objects or arrays", () => {
+    const data = [1, 2, 3, 4, 5, 6];
+    const even = { even: true };
+    const odd = ["odd"];
+    const result = Map.groupBy(data, (num) => (num % 2 === 0 ? even : odd));
+
+    expect(result.size).toBe(2);
+    expect(result.get({ even: true })).toBeUndefined();
+    expect(result.get(["odd"])).toBeUndefined();
+
+    type KeyType = typeof even | typeof odd;
+
+    expect(result).toStrictEqual(
+      new Map<KeyType, number[]>([
```

Comment:
I would guess that `typeof even` gets inferred as `{ even: boolean }` and `typeof odd` gets inferred as `string[]`. If you wanted them to be inferred as something more specific, you can mark the type itself as const, with:

```ts
const even = { even: true } as const;
const odd = ["odd"] as const;
```

That said, was this needed by TypeScript? What happens if we don't specify the type? If we do need it for TypeScript, I'd have a slight preference for inlining the specific type rather than inferring it, since inferring it is a bit circular. Literals are valid types in TypeScript (and what would get inferred if we used `as const`):

```suggestion
    expect(result).toStrictEqual(
      new Map<{ even: true} | ["odd"], number[]>([
```

Reply:
Yes Typescript needed some sort of prodding to accept this without complaint.

Using `as const` and inlining the exact types works and so I've committed that. However, widening the inline types without using `as const` works (without a typescript error at least) as well.